### PR TITLE
Api turn

### PR DESCRIPTION
**Pull request by cnsr**
Basically, how everything works

1. Incoming requests adds `Case` object.
2. If `Case` instance `is_active=True` - a task `send_contact_update` is dispatched.
3. In 14 days, periodic task `perform_nofitications_check` fetches all `Case` instances where `date_end <= date.now()` .
4. All instances are sent to get processed via `send_contact_update` dispatch.

### CODE REVIEW

**Comment by cnsr** [comment]
There's something wrong with either QA image or my docker config.

Ran QA command locally:

```python
root@dcd77a7d0c0d:/app# DJANGO_SETTINGS_MODULE=healthcheck.settings.test python manage.py makemigrations --dry-run | grep 'No changes detected' || (echo 'There are changes which require migrations.' && exit 1)
No changes detected
root@dcd77a7d0c0d:/app#   flake8 ./healthcheck
root@dcd77a7d0c0d:/app#   black --check .
All done! ✨ 🍰 ✨
39 files would be left unchanged.
root@dcd77a7d0c0d:/app#   python manage.py test
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
INFO:users.tests:Created superuser
INFO:users.tests:Logged in with superuser
INFO:users.tests:Created test user
.INFO:users.tests:Created superuser
INFO:users.tests:Logged in with superuser
INFO:users.tests:Created test user without password
...[2020-08-03 12:52:09,310: WARNING/MainProcess] Bad Request: /v1/confirmed_contact/
[2020-08-03 12:52:09,327: WARNING/MainProcess] Bad Request: /v1/confirmed_contact/
[2020-08-03 12:52:09,339: WARNING/MainProcess] Bad Request: /v1/confirmed_contact/
[2020-08-03 12:52:09,349: WARNING/MainProcess] Bad Request: /v1/confirmed_contact/
....
----------------------------------------------------------------------
Ran 8 tests in 7.006s

OK
Destroying test database for alias 'default'...

```

**Comment by rudigiesler** [comment]
The issue in the CI seems to be when running the ` DJANGO_SETTINGS_MODULE=healthcheck.settings.test python manage.py makemigrations --dry-run` command, the error is `KeyError: 'healthcheck'`, which seems to be happening with `mod = importlib.import_module(self.SETTINGS_MODULE)`, so when it's importing the settings.

I think this is related to where you've got things like `CELERY_BROKER_URL = env.str("CELERY_BROKER_URL")` added in the base settings. This syntax means that if `CELERY_BROKER_URL` is not defined, it will raise an error. If you do an `unset CELERY_BROKER_URL` locally, I'm sure you'll get the same error. (Same with all the other config settings that were added in this manner). 

Doing something similar to `CELERY_BROKER_URL = env.str("CELERY_BROKER_URL", "redis://redis:6379/0")` for all those config values will provide a default, which will ensure that things don't fail if that environment variable is missing.

**Comment by cnsr** [comment]
Could the issue be with celery installation? Perhaps, celery fails to load `healthcheck` app (supplied via `-A` flag) in which its config is located?

**Comment by rudigiesler** [comment]
@cnsr I get the same error locally, when running `DJANGO_SETTINGS_MODULE=healthcheck.settings.test python manage.py makemigrations --dry-run`, so I don't think so, no.

I see there's still `TURN_API_KEY`, which doesn't have a default.

**Comment by rudigiesler** [comment]
Running `CELERY_ACCEPT_CONTENT='application/json' TURN_API_KEY=foo DJANGO_SETTINGS_MODULE=healthcheck.settings.test python manage.py makemigrations --dry-run` works fine locally for me, so I think those two settings still need defaults.

But also, it hangs trying to connect to Redis, so I think the default hostname should probably be changed to `localhost`, instead of `redis`, for an easier default.

**Comment by cnsr** [comment]
```sh
celery         |   File "/app/healthcheck/settings/base.py", line 137, in <module>
celery         |     CELERY_ACCEPT_CONTENT = env.list("CELERY_ACCEPT_CONTENT", u"application/json")
celery         |   File "/usr/local/lib/python3.7/site-packages/environ/environ.py", line 179, in list
celery         |     return self.get_value(var, cast=list if not cast else [cast], default=default)
celery         |   File "/usr/local/lib/python3.7/site-packages/environ/environ.py", line 290, in get_value
celery         |     value = self.parse_value(value, cast)
celery         |   File "/usr/local/lib/python3.7/site-packages/environ/environ.py", line 313, in parse_value
celery         |     value = list(map(cast[0], [x for x in value.split(',') if x]))
celery         | TypeError: 'str' object is not callable
```

This is the error i get when the following is set:
```py
CELERY_ACCEPT_CONTENT = env.list("CELERY_ACCEPT_CONTENT", "application/json")
```
When the next config is used, there is no error:
```py
CELERY_ACCEPT_CONTENT = env.list("CELERY_ACCEPT_CONTENT", "") or ["application/json"]
```

Setting defaults to all Celery values and `TURN_API_KEY` did not help.

**Comment by cnsr** [comment]
QA does not even finish right now. I'll look into it tomorrow.

**Comment by cnsr** [comment]
No luck. I have no idea why doesn't this work. 😫

**Comment by cnsr** [comment]
The issue HAS to be with my celery confg.

**Comment by cnsr** [comment]
Another issue could be circular imports.
I had to comment out a line in `Dockerfile`
```yaml
RUN DJANGO_SETTINGS_MODULE=healthcheck.settings.test ./manage.py collectstatic --noinput
```

because my changes to `__init__.py` made it get stuck. 
Mentioned changes are 
```py
from .celery_app import app as celery_app

__all__ = ("celery_app",)
```
which seem to be a must have with celery, commenting them out leads to 
```
celery         | Error: 
celery         | Unable to load celery application.
celery         | Module 'healthcheck' has no attribute 'celery'
celery         | 
celery exited with code 1
```

**Comment by cnsr** [comment]
The main issue were circular imports.
Celery did not see the `healthcheck` package - not sure why, perhaps, the `__init__.py` not being empty was the reason. I found an old project of mine where i had celery and docker configured and looked up the code in it.
Now i just have to figure out what is wrong with `manage.py test`.

**Comment by cnsr** [comment]
Not sure why stuck on `manage.py test`. Tests run fine locally.
Probably due to tests using celery?

also: Current solution is using hardcoded 
```py
CELERY_ACCEPT_CONTENT = ["application/json"]
```
since it does not seem to work otherwise.

**Comment by rudigiesler** [comment]
@cnsr I'm happy with hardcoding that. The only thing we'll need configurable is the broker URL. We don't need to store task results, and hardcoding the content types to JSON will be good for security.

For the error, it seems to be timing out, I'm guessing because there's no Redis in the CI environment. Usually we don't use a full celery testing environment for unit tests. Could we remove the requirement for Redis for running tests by using celery always eager? Right now we use sqlite for unit tests, so we have no external dependancies to be able to run the tests. Would be good to keep it that way.

If not, you can add redis to the github actions config

**Comment by cnsr** [comment]
Does not seem like that helped much.
I'll look into tests - maybe there is an issue with them getting stuck? I've used a very generous 300s for `soft_time_limit` and `time_limit`. I'll try lowering both values to around 15 seconds.

**Comment by rudigiesler** [comment]
@cnsr That could be it. If one of the HTTP requests isn't mocked out correctly, then it could be attempting to call out to an external service. But I don't think that `soft_time_limit` and `time_limit` work when you're running with always eager

**Comment by cnsr** [comment]
Seems like it is, in fact, getting stuck on the test in which celery is being tested - this one:
```
Wed, 05 Aug 2020 14:15:05 GMT
INFO:contacts.test_tasks:Started testing date_start
```

**Comment by cnsr** [comment]
It's actually `task_always_eager` in celery 4 instead of `CELERY_ALWAYS_EAGER`. Will try that now.

**Comment by rudigiesler** [comment]
@cnsr In celery 4, I think it's `CELERY_TASK_ALWAYS_EAGER` then, due to the `app.config_from_object("django.conf:settings", namespace="CELERY")`

**Comment by rudigiesler** [comment]
Looking at the tests, it seems like you've got a `@override_settings(CELERY_ALWAYS_EAGER=True)` on the `setUp`, which will mean that setting will be set during setup, but not during the actual tests.

I would add it to `settings/test.py`, so that it's active for all tests

**Comment by cnsr** [comment]
Ok so the tests i was having issues with are now passing just fine but the tests I was sure were ok are now getting everything stuck.

**Comment by cnsr** [comment]
I am not sure what is happening with `contacts/tests.py` - they seem to be not working after `contacts/test_tasks.py` were configured to work.

File: healthcheck/settings/production.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from healthcheck.settings.base import *  # noqa: F403
 
-SECRET_KEY = env.str("SECRET_KEY")  # noqa: F405
+SECRET_KEY = env.str("SECRET_KEY", "replaceme")  # noqa: F405
```
Comment (rudigiesler) [inline healthcheck/settings/production.py]: I think we should put this back as it was, we don't want to risk running a production instance without a proper secret key, we would rather have it fail to start up

File: healthcheck/settings/base.py [hunk 2]
```diff
@@ -138,3 +122,20 @@
...
 PHONENUMBER_DEFAULT_REGION = "ZA"
+
+# CELERY SETTINGS
+CELERY_BROKER_URL = env.str("CELERY_BROKER_URL", "redis://localhost:6379/0")
+CELERY_RESULT_BACKEND = env.str("CELERY_RESULT_BACKEND", "redis://localhost:6379/0")
+CELERY_ACCEPT_CONTENT = ["application/json"]
+CELERY_TASK_SERIALIZER = "json"
+CELERY_RESULT_SERIALIZER = "json"
+
+# TEMPORARY SETUP - REPLACE WITH @PERIODIC IN PRODUCTION
```
Comment (rudigiesler) [inline healthcheck/settings/base.py]: This is what will be running in production, we can't change files to be different in production.

Does this mean that you'll change this before merging?

Comment (cnsr) [reply healthcheck/settings/base.py]: Yes, I will. I'm going to fix all the other issues and change this to periodic task

File: healthcheck/settings/production.py (hunk 1 above)
Comment (cnsr) [reply healthcheck/settings/production.py]: Got it, will change

File: healthcheck/settings/base.py (hunk 2 above)
Comment (cnsr) [reply healthcheck/settings/base.py]: Changed these to be configurable, made the task periodic.

**Review by rudigiesler — APPROVED** [review]

**Review by rudigiesler — APPROVED** [review]

**Pull request merged**